man/ipescript.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
.TH IPESCRIPT 1 "August 28, 2011"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
ipescript \- Run an Ipe script (written in Lua)
.SH SYNOPSIS
.B ipescript
\fIscript\fP
{ \fIarguments\fP }

.SH DESCRIPTION
.PP
\fBipescript\fP runs an Ipe script written in Lua.  The Ipe bindings
are automatically available in the script.  Ipescript looks for the
script in any of the script directories.

.PP
The \fIscript\fP argument should not contain the .lua extension.

.SH ENVIRONMENT VARIABLES

\fBipescript\fP respects the following environment variables:

.TP
\fBIPELATEXDIR\fP
the directory where \fBipetoipe\fP runs Pdflatex.
.TP
\fBIPELATEXPATH\fP
the directory containing pdflatex and xelatex.  If not set, they are
searched for on the path.
.TP
\fBIPESCRIPTS\fP a list of directories where Ipescript will look for
scripts.  When this variable is not set, Ipe searches first the
current directory, then \fI~/.ipe/scripts\fP, and finally the
system-wide Ipe script directory.  When this variable is set, it
should contain a list of directories, separated by semicolons.  A
single underscore is replaced by the system-wide Ipe script directory.
.TP
\fBIPEDEBUG\fP
set this to 1 for debugging output from \fBipescript\fP.

.SH AUTHOR
Otfried Cheong

.SH REPORTING BUGS
.ad l
Please report bugs using Ipe bugzilla at
.I "http://ipe.otfried.org/bugzilla.html"

.SH SEE ALSO
.ad l
Ipe is documented fully in
.IR "The Ipe Manual" ,
which is available online at
.I "http://ipe.otfried.org/manual/manual.html" .


doc/news.txt
----------------------------------------------------------------------

Ipe 7.2.28

 * Add "sequential text" option.

 * Fix centering of title in presentation stylesheet (bug #464).

----------------------------------------------------------------------

Ipe 7.2.27 (2023-05-05)

 * On Linux, migrate to Qt6 (#457).  Thanks Henri!

 * Change behaviour of transform tools to never switch the current
   selection (#459).

 * Action to deselect everything on current page.

 * Align center/H center/V center now aligns to the page if
   only one object selected (#449).

 * Fix absolute autosave paths (bug #433).

 * Add /opt/homebrew/bin to path on MacOS (#440).

 * Fix missing 'stroke-opacity' and 'opacity' attributes in
   'ipe.dtd' (#67).

 * Fix crash when 'recent_files_lua' ends up empty (bug #455).

 * Fix missing page numbers when there is no text (bug #450).

 * Don't lose group's link action during group edit (bug #456).

----------------------------------------------------------------------

Ipe 7.2.26 (2022-07-16)

 * Allow to store autosave file in same directory as document
   itself.  This is now also the default. (#426)

 * Circular graph edges no longer get broken when moving graph
   nodes (bug #424).

 * Make it possible to remap left_command on MacOS (bug #431).

 * Use PKG_CONFIG environment variable (#429).

----------------------------------------------------------------------

Ipe 7.2.25 (2022-06-19) Denmark Edition

 * Add spell-checking to text entry on Linux.

 * Change presentation style sheet to have 16:9 aspect ratio (#385).

 * Many more configuration options in prefs.canvas_style (#400 &
   #406).

 * Remove box around links in PDF output (#397).

 * Add 'script' text size (#410).

 * Use most-recently used directory in "Save" and "Save As" (#394).

 * Maintain a single custom string on each Ipe object. It's not
   used by Ipe itself, but can be used by external software or
   ipelets. (#409)

 * Add page transition effects from PDF 1.5 (Push, Cover, Uncover,
   Fade).

 * Adjust highlighting text colors in dark theme.

 * Fix crash when compiling with -D_GLIBCXX_ASSERTIONS (#387).

 * Try to simply use UTF-8 in 'runlatex.bat'.  This should work
   at least on Windows-10 (bug #338, second case).

 * Upgrade Freetype in Windows package (fixes bug #411).

 * Show error message when a style sheet mentioned in prefs.styles
   cannot be read (bug #405).

 * Reuse PDF text resources for identical text objects (#412).

 * Fix losing primary selection when deselecting using rectangle drag
   (bug #423).

----------------------------------------------------------------------

Ipe 7.2.24 (2021-04-07)

 * Fix a crash on MacOS 11 El Sur by applying a patch to cairo
   (bug #370).

 * Switch from CAIRO_FILTER_FAST to CAIRO_FILTER_GOOD for decent
   display (essential when filming from the canvas).

 * Make "Select by type/attribute" ipelet work when selecting on an
   absolute color (#380).

 * Fix semi-transparent bitmaps as well as bitmaps with color keys not
   working in Ipepresenter (bug #9).

 * Fix exporting to EPS when part of the drawing lies outside the
   paper (bug #369).

 * Remember online latex service URL in a different file to avoid
   problems when using old and new Ipe versions.

 * Improve calling of external editor on Qt, so that keyboard focus
   is given to the external editor.

 * Raise Ipe window after returning from external editor on Qt, in
   case the editor program is still running.

 * New preference 'freeze_in_latex' to let the Ipe UI freeze during a
   Latex run.

 * New shell script 'iperender-par' that allows to run several
   instances of iperender in parallel (e.g. as part of a parallel
   make) (#375).

 * Add Lua API to access (Latex) dimensions of text object.

 * Silently ignore non-integer grid sizes (#372).

 * On MacOS 11, return to the classic toolbar underneath the title
   bar, because the new unified look leaves most tool buttons off the
   screen (bug #376).

----------------------------------------------------------------------

Ipe 7.2.23 (2020-12-24)

 * Shortcut for pretty display.

 * A maximum on the number of recent files remembered.

 * Improve scaling of toolbar buttons on El Sur, and add labels for
   the toolbar overflow menu.
 
 * Remove sources of libspiro from Ipe releases.

 * Remove doxygen documentation from Ipe release, run doxygen as part
   of the build process.

----------------------------------------------------------------------

Ipe 7.2.22 (2020-12-20)

 * Recent file menu (#322).

 * Update PDF resources on main thread to avoid race condition (bugs
   #354, #359).

 * Can exclude pages from PDF output by adding a layer "NOPDF" to the
   page. The layer does not need to be visible. (#362).  

 * Use a better bitmap checksum that is less likely to cause bitmaps
   to be reordered.  Sort mappings of symbolic attributes to make them
   deterministic (#356).

 * Display Type3 fonts with a unique representation to make it easy to
   diagnose the problem.

----------------------------------------------------------------------

Ipe 7.2.21 (2020-10-26)

 * Fix hanging with message "Running Latex" on Windows (bug #338).

 * Ability to hide the properties panel/layer list from the UI on
   macOS, for use on Zoom etc (#350).
 
 * Remove NoFocus policy from selectors in Qt.  This used to be
   necessary to keep shortcuts working when the selector receives
   focus, but that's no longer necessary (tested in Qt 5.9).
   Fixes selectors not working on Qt 5.15 (bug #341).

 * Fix incorrect update of grid size and angle size selector when
   style sheet is changed (bug #344).

 * Fix crash in 'iperender -pdf' when the document uses
   \includegraphics on a jpg file (bug #342).

 * Fix snapping to midpoint of closing segment of closed paths (bug
   #346).

 * Fix Fifi remaining invisible after switching back from ink mode
   (bug #347).

 * Escape key now closes page/view selectors and sorters on Windows
   (bug #349).


src/ipescript/ipescript.cpp
// --------------------------------------------------------------------
// ipescript.cpp
// --------------------------------------------------------------------
/*

    This file is part of the extensible drawing editor Ipe.
    Copyright (c) 1993-2023 Otfried Cheong

    Ipe is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    As a special exception, you have permission to link Ipe with the
    CGAL library and distribute executables, as long as you follow the
    requirements of the Gnu General Public License in regard to all of
    the software in the executable aside from CGAL.

    Ipe is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
    License for more details.

    You should have received a copy of the GNU General Public License
    along with Ipe; if not, you can find it at
    "http://www.gnu.org/copyleft/gpl.html", or write to the Free
    Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*/

#include "ipebase.h"
#include "ipelua.h"

#include <cstdio>
#include <cstdlib>

using namespace ipe;
using namespace ipelua;

#ifdef WIN32
#define IPEPATHSEP (';')
#else
#define IPEPATHSEP (':')
#endif

// --------------------------------------------------------------------

static int traceback (lua_State *L)
{
  if (!lua_isstring(L, 1))  /* 'message' not a string? */
    return 1;  /* keep it intact */
  lua_rawgeti(L, LUA_REGISTRYINDEX, LUA_RIDX_GLOBALS);
  lua_getfield(L, -1, "debug");
  if (!lua_istable(L, -1)) {
    lua_pop(L, 2);
    return 1;
  }
  lua_getfield(L, -1, "traceback");
  if (!lua_isfunction(L, -1)) {
    lua_pop(L, 3);
    return 1;
  }
  lua_pushvalue(L, 1);             // pass error message
  lua_pushinteger(L, 2);           // skip this function and traceback
  luacall(L, 2, 1);                // call debug.traceback
  return 1;
}

static void setup_config(lua_State *L, const char *var, const char *conf)
{
#ifdef IPEBUNDLE
  push_string(L, Platform::ipeDir(conf));
#else
  lua_pushstring(L, conf);
#endif
  lua_setfield(L, -2, var);
}

// --------------------------------------------------------------------

static lua_State *setup_lua()
{
  lua_State *L = luaL_newstate();
  luaL_openlibs(L);
  luaopen_ipe(L);
  return L;
}

static void setup_globals(lua_State *L)
{
  lua_getglobal(L, "package");
  const char *pscripts = getenv("IPESCRIPTS");
  if (pscripts) {
    String scripts(pscripts);
    String s;
    int i = 0;
    while (i < scripts.size()) {
      int j = i;
      while (i < scripts.size() && scripts[i] != IPEPATHSEP)
	i += 1;
      String d = scripts.substr(j, i-j);
      if (d == "_")
#ifdef IPEBUNDLE
	d = Platform::ipeDir("scripts", nullptr);
#else
	d = IPESCRIPTDIR;
#endif
      d += "/?.lua";
      if (!s.empty())
	s += ';';
      s += d;
      i += 1;
    }
    ipeDebug("package.path = %s", s.z());
    lua_pushstring(L, s.z());
  } else {
    String s = "./?.lua;";
#ifndef WIN32
    const char *home = getenv("HOME");
    if (home) {
      s += home;
      s += "/.ipe/scripts/?.lua;";
#ifdef __APPLE__
      s += home;
      s += "/Library/Ipe/Scripts/?.lua;";
#endif
    }
#endif
#ifdef IPEBUNDLE
    s += Platform::ipeDir("scripts", "?.lua");
#else
    s += IPESCRIPTDIR "/?.lua";
#endif
    ipeDebug("package.path = %s", s.z());
    lua_pushstring(L, s.z());
  }
  lua_setfield(L, -2, "path");

  lua_newtable(L);  // config table
#if defined(WIN32)
  lua_pushliteral(L, "win");
#elif defined(__APPLE__)
  lua_pushliteral(L, "apple");
#else
  lua_pushliteral(L, "unix");
#endif
  lua_setfield(L, -2, "platform");

#ifdef IPEBUNDLE
  setup_config(L, "system_styles", "styles");
#else
  setup_config(L, "system_styles", IPESTYLEDIR);
#endif
  push_string(L, Platform::latexDirectory());
  lua_setfield(L, -2, "latexdir");

  lua_pushfstring(L, "Ipe %d.%d.%d",
		  IPELIB_VERSION / 10000,
		  (IPELIB_VERSION / 100) % 100,
		  IPELIB_VERSION % 100);
  lua_setfield(L, -2, "version");

  lua_setglobal(L, "config");
}

// --------------------------------------------------------------------

static void usage()
{
  fprintf(stderr,
	  "Usage: ipescript <script> { <arguments> }\n"
	  "Ipescript runs a script from your scripts directories with\n"
	  "the given arguments.\n"
	  "Do not include the .lua extension in the script name.\n");
  exit(1);
}

int main(int argc, char *argv[])
{
  Platform::initLib(IPELIB_VERSION);

  if (argc < 2)
    usage();

  String s = "require \"";
  s += argv[1];
  s += "\"";

  lua_State *L = setup_lua();

  // create table with arguments
  lua_createtable(L, 0, argc - 2);
  for (int i = 2; i < argc; ++i) {
    lua_pushstring(L, argv[i]);
    lua_rawseti(L, -2, i - 1);
  }
  lua_setglobal(L, "argv");

  setup_globals(L);

  // run script
  lua_pushcfunction(L, traceback);
  assert(luaL_loadstring(L, s.z()) == 0);

  if (lua_pcallk(L, 0, 0, -2, 0, nullptr)) {
    const char *errmsg = lua_tolstring(L, -1, nullptr);
    fprintf(stderr, "%s\n", errmsg);
  }

  lua_close(L);
  return 0;
}

// --------------------------------------------------------------------
